cpp/libcudf_kafka/kafka/src/kafka_consumer.cpp

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kafka_consumer.hpp"
+#include <librdkafka/rdkafkacpp.h>
+
+namespace cudf {
+namespace io {
+namespace external {
+namespace kafka {
+
+kafka_consumer::kafka_consumer(std::map<std::string, std::string> configs,
+                               std::string topic_name,
+                               int partition,
+                               int64_t start_offset,
+                               int64_t end_offset,
+                               int batch_timeout,
+                               std::string delimiter)
+  : topic_name(topic_name),
+    partition(partition),
+    start_offset(start_offset),
+    end_offset(end_offset),
+    batch_timeout(batch_timeout),
+    delimiter(delimiter)
+{
+  kafka_conf = std::unique_ptr<RdKafka::Conf>(RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL));
+
+  // Set configs and check for invalids
+  RdKafka::Conf::ConfResult conf_res;
+  std::string errstr;    // Textual representation of Error
+  std::string conf_val;  // String value of a RDKafka configuration request
+
+  for (auto const &x : configs) {
+    conf_res = kafka_conf->set(x.first, x.second, errstr);
+    if (conf_res == RdKafka::Conf::ConfResult::CONF_UNKNOWN) {
+      CUDF_FAIL("'" + x.first + ", is an invalid librdkafka configuration property");
+    } else if (conf_res == RdKafka::Conf::ConfResult::CONF_INVALID) {
+      CUDF_FAIL("'" + x.second +
+                "' contains an invalid configuration value for librdkafka property '" + x.first +
+                "'");
+    }
+  }
+
+  // Kafka 0.9 > requires group.id in the configuration
+  conf_res = kafka_conf->get("group.id", conf_val);
+  if (conf_res == RdKafka::Conf::ConfResult::CONF_UNKNOWN) {
+    CUDF_FAIL(
+      "Kafka `group.id` was not supplied in its configuration and is required for operation");
+  }
+
+  consumer = std::unique_ptr<RdKafka::KafkaConsumer>(
+    RdKafka::KafkaConsumer::create(kafka_conf.get(), errstr));
+
+  // Pre fill the local buffer with messages so the datasource->size() invocation
+  // will return a valid size.
+  consume_to_buffer();
+}
+
+std::unique_ptr<cudf::io::datasource::buffer> kafka_consumer::host_read(size_t offset, size_t size)
+{
+  if (offset < buffer.size()) { size = std::min(size, buffer.size() - offset); }
```

Comment:
You should return an empty buffer if offset is outside of the valid range. 
Then the size clamp does not need to be conditionally done.